llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -265,6 +281,76 @@ void collectCompositeElementsInfoRecursive(
+    for (size_t I = 0; I < ArrTy->getNumElements(); ++I) {
+      Constant *El = cast<Constant>(C->getOperand(I));
+      // TODO: this is a spot for potential optimization: for arrays we could
+      // just make a single recursive call here and use it to populate
+      // DefaultValues in a loop.
+      collectCompositeElementsDefaultValuesRecursive(M, El, Index, Offset,
+                                                     DefaultValues);
+    }
+  } else if (auto *StructTy = dyn_cast<StructType>(Ty)) {
+    const StructLayout *SL = M.getDataLayout().getStructLayout(StructTy);
+    for (size_t I = 0, E = StructTy->getNumElements(); I < E; ++I) {
+      Constant *El = cast<Constant>(C->getOperand(I));
+      // When handling elements of a structure, we do not use manually
+      // calculated offsets (which are sum of sizes of all previously
+      // encountered elements), but instead rely on data provided for us by
+      // DataLayout, because the structure can be unpacked, i.e. padded in
+      // order to ensure particular alignment of its elements.
+      unsigned LocalOffset = Offset + SL->getElementOffset(I);
+
+      // If there was some alignment, fill the data between values with zeros.
+      while (LocalOffset != DefaultValues.size())
+        DefaultValues.push_back(0);
+
+      collectCompositeElementsDefaultValuesRecursive(M, El, Index, LocalOffset,
+                                                     DefaultValues);
+    }
+    // Update "global" offset according to the total size of a handled struct
+    // type.
+    Offset += SL->getSizeInBytes();
+  } else if (auto *VecTy = dyn_cast<FixedVectorType>(Ty)) {
+    for (size_t I = 0; I < VecTy->getNumElements(); ++I) {
+      Constant *El = cast<Constant>(C->getOperand(I));
+      // TODO: this is a spot for potential optimization: for vectors we could
+      // just make a single recursive call here and use it to populate
+      // DefaultValues in a loop.
+      collectCompositeElementsDefaultValuesRecursive(M, El, Index, Offset,
+                                                     DefaultValues);
+    }
+  } else { // Assume that we encountered some scalar element
+    int NumBytes = Ty->getScalarSizeInBits() / CHAR_BIT;
+    char *CharPtr;
+
+    if (auto IntConst = dyn_cast<ConstantInt>(C)) {
+      auto Val = IntConst->getValue().getZExtValue();
+      CharPtr = reinterpret_cast<char *>(&Val);
+    } else if (auto FPConst = dyn_cast<ConstantFP>(C)) {
+      auto Val = FPConst->getValue();
+
+      if (NumBytes == 4) {
+        float v = Val.convertToFloat();
+        CharPtr = reinterpret_cast<char *>(&v);
+      } else if (NumBytes == 8) {
+        double v = Val.convertToDouble();
+        CharPtr = reinterpret_cast<char *>(&v);
+      }
+    }
+    std::copy_n(CharPtr, NumBytes, std::back_inserter(DefaultValues));
+    Index++;
+    Offset += Ty->getPrimitiveSizeInBits() / 8 +
+              (Ty->getPrimitiveSizeInBits() % 8 != 0);
```

Comment:
Since you are appending to a vector and not inserting into a pre-allocated vector, you probably don't need this argument anymore

Reply:
`Offset` argument is actually used for calculations in `StructTy` - it helps to fill the buffer with the correct alignment:
https://github.com/intel/llvm/pull/3666/files#diff-f9a582625f70ea58ef88b7713c591e8187629cca507b424557a3e4664e23848fR308-R312